Fix empty Repeater rows and default values

kienstra opened the pull request:
# WIP

## Before
* On adding empty rows, they did not appear at all in the editor
* This looks to be from the way `<ServerSideRender>` [handles empty {}](https://github.com/WordPress/gutenberg/blob/e2490612decf2f0ac614d6ee9f6d1269d0bfc615/packages/server-side-render/src/server-side-render.js#L22)
* Default values were only applied to the first row

## After 
* This adds a workaround for that, setting the new row to `{ '': '' }`
* This still isn't ideal, and I should keep an eye out for a better approach
* Default values are now applied to all repeater rows, even for completely empty rows

Fixes #393

kienstra committed:
Add a workaround for empty rows not appearing in the Repeater

Before, on adding empty rows,
they did not appear at all in the editor.
This looks to be from the way <ServerSideRender>
handles empty {}.
So this adds a workaround for that, setting
the new row to { '': '' }.
This still isn't ideal, and I should keep an eye
out for a better approach.

kienstra commented on js/blocks/components/edit.js:
```diff
@@ -7,7 +7,7 @@ const { Fragment } = wp.element;
  * Internal dependencies
  */
-import { AdvancedControls, BlockLabInspector, Fields, FormControls } from './';
+import { AdvancedControls, BlockLabInspector, FormControls } from './';
```
I should have removed `Fields` earlier, it's not used in the file.

kienstra commented on js/blocks/controls/repeater.js:
```diff
@@ -10,14 +10,28 @@ const { BaseControl, IconButton } = wp.components;
 
 const BlockLabRepeaterControl = ( props ) => {
-	const { field, getValue, onChange, parentBlock, parentBlockProps } = props;
+	const { field, onChange, parentBlock, parentBlockProps } = props;
```
I should have removed `getValue` earlier, it's not used in the file.

kienstra commented on js/blocks/controls/repeater.js:
```diff
@@ -10,14 +10,28 @@ const { BaseControl, IconButton } = wp.components;
+	 * @see https://github.com/getblocklab/block-lab/issues/393
+	 */
+	const addEmptyRow = () => {
+		const withAddedRow = rows.concat( { '': '' } );
```
This isn't ideal, but it's worked so far in brief testing. I'll keep testing this.

kienstra committed:
Remove needless parentheses

isset() does not need to be wrapped
in ().

lukecarbis commented:
@kienstra I can confirm that this works great for adding and saving new blank rows.

I'm still experiencing this blank rows bug, I wonder if it's related:

![2019-09-03 06-53-19 2019-09-03 06_59_04](https://user-images.githubusercontent.com/1097667/64133292-6f19a780-ce18-11e9-9e14-82ec40cbe04c.gif)

Happy to create a new issue / PR to resolve this if it's sufficiently different, but I thought the issues might be tied together in some way.

kienstra commented:
>I'm still experiencing this blank rows bug, I wonder if it's related:

Ah, interesting. I'll look at that.

kienstra commented:
So far, it looks like this [problem](https://github.com/getblocklab/block-lab/pull/409#issuecomment-527242700) only exists with fields that have `<TextControl>`, like Text and Email. Though I'm not sure they're causing this.

kienstra committed:
Address an issue with moving rows, where the moved row is duplicated

As Luke pointed out,
when moving a row up or down,
if the row that is exchanged has empty values,
they will be replaced with the moved row's values.
But this might not be a complete solution.

kienstra commented:
**Checkbox Issue**

Hi @lukecarbis,
Good talking to you 😄 

I also saw the issue with the checkbox you mentioned, where it appears to be 'half-checked.'

https://github.com/getblocklab/block-lab/pull/409/commits/c55a103ca1c545a47ed20b3b695717c1c5de4839 fixed it in my testing, but let me know if you still see the issue.

Feel free to wait until a code review if you'd like. 

Thanks, Luke!

kienstra committed:
Remove extra empty line that I added.

lukecarbis commented on js/blocks/components/repeater-rows.js:
```diff
@@ -129,26 +129,7 @@ import { Fields } from './';
-					1
-				)[0]
-			);
+			[ rows[ from ], rows[ to ] ] = [ rows[ to ], rows[ from ] ]
```
Wow – this is so so obvious. Can't believe how much I over-engineered this! Pickup of the week!

kienstra replied:
Haha, I didn't know about that either. I just happened upon it from StackOverflow.

kienstra replied:
The previous solution actually worked fine.

lukecarbis commented on js/blocks/controls/email.js:
```diff
@@ -4,6 +4,8 @@ const { TextControl } = wp.components;
 const BlockLabEmailControl = ( props ) => {
 	const { field, getValue, onChange } = props;
+	const initialValue = getValue( props );
+	const value = 'undefined' !== typeof initialValue ? initialValue : field.default;
```
Great solution. Really glad to get this console error sorted.

kienstra replied:
Thanks!

kienstra replied:
Thanks!

lukecarbis reviewed (approved):
I messed around for quite a while with new blank rows, saved blank rows, moving blank rows around, etc. As far as I can see, this PR solves all issues with blank rows. Great work! Thank you!

kienstra commented:
Hi @lukecarbis, thanks a lot for reviewing and testing this. I'm going to merge it now if that's alright.
